### ISSUE TITLE
Don't store NoopFactoryMachines

### ISSUE THREAD

**Issue by a-peyrard** [restx/restx#151]
Currently when we use alternative components (or conditional ones), a FactoryMachine producing a FactoryMachine for the component is created (for example the generated source: `restx.specs.InfinirestRestxSpecRepositoryFactoryMachine`). 
During the construction of the Factory, those FactoryMachines will be built and will produced machines stored in the Factory (in the `machines` list) and boxes will be stored in a temporary warehouse. In any cases a FactoryMachine is produced even if the condition is not fulfil, the machine stored in that case is the NoopFactoryMachine instance.

So if there is a lot of alternative components or conditional components, there will be some useless machines queried by the factory.
Xavier point me out another problem with that, it is the use of a temporary warehouse during the construction. Dependencies of those factory machines will be built, and stored in this temporary warehouse, so if they are used after that, they will be built again.

I tried to implement a very basic solution in that branch : https://github.com/a-peyrard/restx/commits/dont-store-factory-machines
Changes are:
- No temporary warehouse is used, the final one is used during the construction of the FactoryMachine
- When the factory stores the new produced machine, it ignores the NoopFactoryMachine instances.
- DisposableBox are no more stored in the warehouse, this part is done there: https://github.com/a-peyrard/restx/commit/ad9a8b5679837742371d335f5d75216796e08ec7 
I'm not a big fan of this, Xavier proposed to checkin the box and to remove it on the checkout if the box is empty. It implies to change the ComponentBox interface to add a method permitting to know is the box is empty or not. Not a big fan of this solution either.

I also tried to think about this problem in a different way. The problem is the `buildAndStore` method, maybe it could be split into two part, the construction, and the storage. Like this `buildFactoryMachines` will only use the construction method. And anyway storage method will ignore disposable boxes. 
It was a thought of a night, not sure its doable though. And I already know Xavier prefer to keep the responsibility of storage and to reuse or not a component, at warehouse and boxes level.  

What do you think about this ?

**Comment by a-peyrard** [comment]
I just fall again on a problem described here.

We have a main factory, where a component initialising the database is created (its the hibernate SessionFactory), this component has to be created only once.
We have several sub-factories linked to the main factory thanks to warehouse providers. Until today everything was fine. But we decided to add a customizer in our sub-factories which is using the SessionFactory component. And during the construction of the customizers, the factory does not have the link to the main warehouse provider, it is using a temporary warehouse. So the SessionFactory is created once again in every sub-factories.

Xavier when we discussed about this subject several month ago, you told me that you had this kind of problem for factory machines of factory machines. Did you had some time to think about a solution for this problem ?
 
In order for our application to work, we are not using DI in the customizer anymore. But this is a pity to have to use a trick like this.

**Comment by a-peyrard** [comment]
I just faced another similar problem. I created a customizer which requires a component A, so the component A is created during the initialisation of the factory, and stored in the temp warehouse.
And later on when the component A is queried, it is created a second time and stored in the production warehouse.

So it breaks singletoness of component...

Any idea ? The more I see, the more I think we shouldn't use a temp warehouse.

**Comment by xhanin** [comment]
Yes I see the point, and I tend to think you're right, we shouldn't use a
temporary warehouse. I'm trying to think about use cases where that would
cause a problem, but so far it's not clear. The only problem I see is that
the components placed in the warehouse during the factory initialization
process won't have the benefit of the fully initialized factory. For
instance, if you have a Machine producing other Machines which in turn
produce a customizer, that customizer won't be applied to components
created during the initialisation of the components required to produce the
Machine. But that sound fair enough. We need to think that it means that
these components won't be hot reloadable.This probably include the settings
/ named strings which are most of the time needed during that phase. Except
if we make a kind of exception for that, I remember that you did something
for that in one of your use case but I'm not sure.

Otherwise, we could either have a general setting to indicate which mode to
use : reuse or not the warehouse during the init process. Or annotate the
components produced to indicate if they need to be true singletons, and
have a single shared warehouse hosting all "true" singletons that we would
always link to any other warehouse. More flexible but also more complex.
And in that case I don't know if we would store these singletons
dependencies also, or clean them so that they can be hot reloaded.

What do you think?



On Thu, Jul 23, 2015 at 12:10 PM, a-peyrard <notifications@github.com>
wrote:

> I just faced another similar problem. I created a customizer which
> requires a component A, so the component A is created during the
> initialisation of the factory, and stored in the temp warehouse.
> And later on when the component A is queried, it is created a second time
> and stored in the production warehouse.
>
> So it breaks singletoness of component...
>
> Any idea ? The more I see, the more I think we shouldn't use a temp
> warehouse.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/restx/restx/issues/151#issuecomment-124043184>.
>



-- 
Xavier Hanin
Lead architect at 4SH France - http://www.4sh.fr/
RESTX creator and lead developer - http://restx.io/
BordeauxJUG creator - http://www.bordeauxjug.org/
Apache Ivy Creator - http://ant.apache.org/ivy/

**Comment by a-peyrard** [comment]
The storage of `NoopFactoryMachine` might be solve by the issue #213, but the reuse of the warehouse still need to be addressed.
